lesson25/index.js

```diff
@@ -0,0 +1,124 @@
+const submitButton = document.getElementById("js-submit-button");
+const checkBox = document.querySelector(".js-check-box");
+const overLay = document.getElementById("js-overlay");
+const closeButton = document.getElementById("js-modal-close");
+const rules = document.getElementById("js-rules");
+const body = document.querySelector("body");
+const form = document.getElementById("js-register-form");
+
+const closeModal = () => {
+  body.classList.remove("modal-open");
+};
+
+const openModal = () => {
+  body.classList.add("modal-open");
+};
+
+rules.addEventListener("click", openModal);
+closeButton.addEventListener("click", closeModal);
+overLay.addEventListener("click", closeModal);
+
+const scrollInModalContent = new IntersectionObserver(
+  ([entry]) => {
+    if (entry.isIntersecting) {
+      checkBox.checked = true;
+      checkBox.disabled = false;
+      const event = new Event("input", { bubbles: true });
+      checkBox.dispatchEvent(event);
+      scrollInModalContent.disconnect();
+    }
+  },
+  { root: modalContent }
+);
+scrollInModalContent.observe(modalContent.lastElementChild);
+
+const constraint = {
+  name: {
+    validation: () => {
+      const value = document.getElementById("name").value;
+      return isLimitTextLength(value, 16);
+    },
+    invalidMessage: "ユーザー名は15文字以下にしてください。"
+  },
+  email: {
+    validation: () => {
+      const value = document.getElementById("email").value;
+      const reg = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+\.[A-Za-z]+$/g;
+      return isValidInRegex(reg, value);
+    },
+    invalidMessage: "メールアドレスの形式になっていません"
+  },
+  password: {
+    validation: () => {
+      const value = document.getElementById("password").value;
+      const reg = /^(?=.*[A-Z])(?=.*[a-z])(?=.*[0-9])[a-zA-Z0-9]{8,}$/g;
+      return isValidInRegex(reg, value);
+    },
+    invalidMessage: "8文字以上の大小の英数字を交ぜたものにしてください。"
+  }
+};
+
+const isBlankInInput = (value) => value.trim() === "";
+
+const isLimitTextLength = (value, limit) => value.length < limit;
+
+const isValidInRegex = (constraint, value) => constraint.test(value);
+
+const addValidClassName = (target) => {
+  const parent = target.parentElement;
+  parent.classList.add("valid");
+};
+
+const removeInvalidMessage = (target) => {
+  target.parentElement.classList.remove("invalid");
+  const errorMessage = target.parentElement.querySelector(".invalid-message");
+  errorMessage && errorMessage.remove();
+};
+
+const removeValidClassName = (target) => {
+  target.parentElement.classList.remove("valid");
+};
+
+const addValidationMessage = (field) => {
+  if (isBlankInInput(field.value)) {
+    removeValidClassName(field);
+    addInvalidMessage(field, "未入力です");
+    return;
+  } else if (!constraint[field.id].validation()) {
+    removeValidClassName(field);
+    addInvalidMessage(field, constraint[field.id].invalidMessage);
+    return;
+  } else {
+    removeInvalidMessage(field);
+    addValidClassName(field);
+  }
+};
```

Comment:
IMO
I think using early returns will improve the outlook.
I also thought it would be easier to read the if statement if it branched with truthy values as much as possible.
[早期リターンを書こう](https://zenn.dev/media_engine/articles/early_return)
```suggestion
const addValidationMessage = (field) => {
  if (isBlankInInput(field.value)) {
    removeValidClassName(field);
    addInvalidMessage(field, "未入力です");
    return;
  }
  if (constraint[field.id].validation()) {
    removeInvalidMessage(field);
    addValidClassName(field);
    return;
  }
  removeValidClassName(field);
  addInvalidMessage(field, constraint[field.id].invalidMessage);
};
```

Reply:
I thought the UI for form validation was very nice.

Reply:
Thanks for the review.
I found the if statement complicated.
The use of early returns gave me a better perspective on processing for conditions.

Also, the condition was changed to truth value.

[30c8b81](https://github.com/sae-github/JS_lesson/pull/29/commits/30c8b81259c74ef5ba8f1a112e1a84e7c32b0ae9)